File: src/main/scala/slick/adapter/SlickJdbcProvider.scala
package slick.adapter

import java.util.concurrent.CompletionStage

import com.typesafe.config.Config
import slick.adapter.JdbcSession.ConnectionFunction
import slick.jdbc.{JdbcBackend, SimpleJdbcAction}

import scala.compat.java8.FutureConverters._

/**
  * This class provides access to Slick managed DB connection pool.
  * It's an adaptor to the JdbcSession Java interface.
  */
class SlickJdbcProvider(config: Config) extends JdbcSession with AutoCloseable {

  private val db = JdbcBackend.createDatabase(config, "")

  private val SimpleDBIO = SimpleJdbcAction

  override def withConnection[T](block: ConnectionFunction[T]): CompletionStage[T] = {
    db.run {
      SimpleDBIO { ctx =>
        block(ctx.connection)
      }
    }.toJava
  }

  override def close(): Unit = {
    db.close()
  }
}

File: src/main/java/slick/adapter/javaapi/SlickJdbcProvider.java
package slick.adapter.javaapi;


import java.sql.SQLException;
import java.util.concurrent.CompletionStage;

import com.typesafe.config.Config;
import slick.adapter.JdbcSession;
import slick.jdbc.JdbcBackend;
import slick.jdbc.JdbcBackend$;
import slick.jdbc.SimpleJdbcAction;

import scala.compat.java8.FutureConverters;

public class SlickJdbcProvider implements JdbcSession, AutoCloseable {

    private final Config config;

    JdbcBackend.DatabaseDef db;

    public SlickJdbcProvider(Config config) {
        this.config = config;

        this.db = JdbcBackend$.MODULE$.createDatabase(config, "");
    }

    @Override
    public void close() {
        db.close();
    }

    @Override
    public <T> CompletionStage<T> withConnection(ConnectionFunction<T> block) {
        SimpleJdbcAction simpleDBIO = new SimpleJdbcAction<>(
                (JdbcBackend.JdbcActionContext ctx) -> {
                        try {
                            return block.apply(ctx.connection());
                        } catch (SQLException e) {
                            throw new RuntimeException(e);
                        }
                }
        );
        return FutureConverters.toJava(db.run(simpleDBIO));
    }


}
